src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +71,95 @@ def install_given_reqs(
             ', '.join(to_install.keys()),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = collections.OrderedDict({name: None for name in to_install})
+
+    install_args = [install_options, global_options, dict(
+        root=root, home=home, prefix=prefix,
+        warn_script_location=warn_script_location,
+        use_user_site=use_user_site, pycompile=pycompile)]
 
     with indent_log():
-        for req_name, requirement in to_install.items():
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', req_name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
-            else:
-                uninstalled_pathset = None
-
-            try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    root=root,
-                    home=home,
-                    prefix=prefix,
-                    warn_script_location=warn_script_location,
-                    use_user_site=use_user_site,
-                    pycompile=pycompile,
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
-
-    return installed
+        # first try to install in parallel
+        installed_pool = _safe_pool_map(
+            partial(_single_install, install_args, in_subprocess=True),
+            list(to_install.values()))
+        if installed_pool:
+            installed = collections.OrderedDict(
+                zip(list(to_install.keys()), installed_pool))
+
+        for name, requirement in to_install.items():
+            if installed[name] is None:
+                installed_req = _single_install(
+                    install_args, requirement, in_subprocess=False)
+                installed[name] = installed_req  # type: ignore
+            elif isinstance(installed[name], BaseException):
+                raise installed[name]   # type: ignore
+
+    return [
+        i for i in installed.values()  # type: ignore
+        if isinstance(i, InstallationResult)
+    ]
+
+
+def _safe_pool_map(
+        func,               # type: Callable[[Any], Any]
+        iterable,           # type: Iterable[Any]
+):
+    # type: (...) -> Optional[List[Any]]
+    """
+    Safe call to Pool map, if Pool is not available return None
+    """
+    if not iterable:
+        return None
```

Comment:
We want to return `None` on an empty iterable. But `bool(iterator)` is always True even if the iterator yields nothing. So we probably need to rewrite this (and the surrounding code calling it).

Reply:
I'm confused.

```
>>> a=[]
>>> b=bool(a)
>>> b
False
```

Reply:
Actually the entire function is now redundant `_map_multiprocess_ordered` will take care of it, refactoring.

Reply:
@LukeIreland1 The thing is, if this is an iterator (not a list) then checking boolean on iterator always returns True :)
Anyhow I removed it since the `_map_multiprocess_ordered` should take care of it as it executes Pool.map

Reply:
> @LukeIreland1 The thing is, if this is an iterator (not a list) then checking boolean on iterator always returns True :)
> Anyhow I removed it since the `_map_multiprocess_ordered` should take care of it as it executes Pool.map

Ah, my bad. Is a list an iterable? I think I got mixed up between iterable and iterator here.

Reply:
[Python’s definition of iterable and iterator.](https://docs.python.org/3/glossary.html#term-iterable)